### ISSUE TITLE
Clicking tableView column header

### ISSUE THREAD

**Issue by protektik** [TestFX/TestFX#541]
I was using version 4.0.4-alpha where, when i done this:
clickOn(tableColumnId);
it clicked on tableColumn header, which was expected behavior. Now I moved to 4.0.11-alpha and the same code:
clickOn(tableColumnId);
clicks on specified column, but somewhere in the data area. Is there a way to click on tableColumn header in 4.0.11-alpha?

## Specifications

  - Version: 4.0.11-alpha
  - Platform: windows 7

**Comment by brcolow** [comment]
Hmmm that's an interesting regression. Not sure how any of the changes could cause that. Do you have a simple test case that shows the expected vs. actual behavior?

**Comment by brcolow** [comment]
On `4.0.12-alpha` calling `clickOn("#tableColumn")` clicks on the first row, for me. I will investigate what changed from `4.0.4-alpha` w.r.t. this issue.

**Comment by brcolow** [comment]
Something is different between the bounds returned by `4.0.4-alpha` and `master`:

These print statements can be seen in https://github.com/brcolow/TestFX/tree/issue-541-2

```
master:
    point locator: org.testfx.service.locator.impl.PointLocatorImpl@13842bad
    pointPosition: CENTER
    PointQueryUtils#computePositionFactors, returning: Point2D [x = 0.5, y = 0.5]
    boundsInSceneFor: TableColumn$1$1[id=tableColumn0, styleClass=cell indexed-cell table-cell table-column]'null'
    sceneBounds: BoundingBox [minX:1.0, minY:234.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:58.0, maxY:263.0, maxZ:0.0]
    Returning: BoundingBox [minX:1597.0, minY:581.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:1654.0, maxY:610.0, maxZ:0.0]
    pointQuery: CallableBoundsPointQuery [position = Point2D [x = 0.5, y = 0.5], offset = Point2D [x = 0.0, y = 0.0], bounds = BoundingBox [minX:1597.0, minY:581.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:1654.0, maxY:610.0, maxZ:0.0]]
    boundsInSceneFor: TableColumn$1$1[id=tableColumn0, styleClass=cell indexed-cell table-cell table-column]'null'
    sceneBounds: BoundingBox [minX:1.0, minY:234.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:58.0, maxY:263.0, maxZ:0.0]
    Returning: BoundingBox [minX:1597.0, minY:581.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:1654.0, maxY:610.0, maxZ:0.0]
    bounds: BoundingBox [minX:1597.0, minY:581.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:1654.0, maxY:610.0, maxZ:0.0]
    positionFactors: Point2D [x = 0.5, y = 0.5]
    PointQueryUtils#atPositionFactors returning: Point2D [x = 1625.5, y = 595.5]
    BoundsPointQuery#query.point: Point2D [x = 1625.5, y = 595.5]
    BoundsPointQuery#query.offet: Point2D [x = 0.0, y = 0.0]
    boundsInSceneFor: TableColumn$1$1[id=tableColumn0, styleClass=cell indexed-cell table-cell table-column]'null'
    sceneBounds: BoundingBox [minX:1.0, minY:234.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:58.0, maxY:263.0, maxZ:0.0]
    Returning: BoundingBox [minX:1597.0, minY:581.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:1654.0, maxY:610.0, maxZ:0.0]
    bounds: BoundingBox [minX:1597.0, minY:581.0, minZ:0.0, width:57.0, height:29.0, depth:0.0, maxX:1654.0, maxY:610.0, maxZ:0.0]
    positionFactors: Point2D [x = 0.5, y = 0.5]
    PointQueryUtils#atPositionFactors returning: Point2D [x = 1625.5, y = 595.5]
    BoundsPointQuery#query.point: Point2D [x = 1625.5, y = 595.5]
    BoundsPointQuery#query.offet: Point2D [x = 0.0, y = 0.0]
```

v4.0.4-alpha:

```
    point locator: org.testfx.service.locator.impl.PointLocatorImpl@6fad4e8b
    pointPosition: CENTER
    PointQueryUtils#computePositionFactors, returning: Point2D [x = 0.5, y = 0.5]
    boundsInSceneFor: TableColumnHeader[id=tableColumn0, styleClass=column-header table-column]
    sceneBounds: BoundingBox [minX:1.0, minY:1.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:57.884765625, maxY:31.0, maxZ:0.0]
    Returning: BoundingBox [minX:1597.0, minY:348.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:1653.884765625, maxY:378.0, maxZ:0.0]
    pointQuery: CallableBoundsPointQuery [position = Point2D [x = 0.5, y = 0.5], offset = Point2D [x = 0.0, y = 0.0], bounds = BoundingBox [minX:1597.0, minY:348.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:1653.884765625, maxY:378.0, maxZ:0.0]]
    boundsInSceneFor: TableColumnHeader[id=tableColumn0, styleClass=column-header table-column]
    sceneBounds: BoundingBox [minX:1.0, minY:1.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:57.884765625, maxY:31.0, maxZ:0.0]
    Returning: BoundingBox [minX:1597.0, minY:348.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:1653.884765625, maxY:378.0, maxZ:0.0]
    bounds: BoundingBox [minX:1597.0, minY:348.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:1653.884765625, maxY:378.0, maxZ:0.0]
    positionFactors: Point2D [x = 0.5, y = 0.5]
    PointQueryUtils#atPositionFactors returning: Point2D [x = 1625.4423828125, y = 363.0]
    BoundsPointQuery#query.point: Point2D [x = 1625.4423828125, y = 363.0]
    BoundsPointQuery#query.offet: Point2D [x = 0.0, y = 0.0]
    boundsInSceneFor: TableColumnHeader[id=tableColumn0, styleClass=column-header table-column]
    sceneBounds: BoundingBox [minX:1.0, minY:1.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:57.884765625, maxY:31.0, maxZ:0.0]
    Returning: BoundingBox [minX:1597.0, minY:348.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:1653.884765625, maxY:378.0, maxZ:0.0]
    bounds: BoundingBox [minX:1597.0, minY:348.0, minZ:0.0, width:56.884765625, height:30.0, depth:0.0, maxX:1653.884765625, maxY:378.0, maxZ:0.0]
    positionFactors: Point2D [x = 0.5, y = 0.5]
    PointQueryUtils#atPositionFactors returning: Point2D [x = 1625.4423828125, y = 363.0]
    BoundsPointQuery#query.point: Point2D [x = 1625.4423828125, y = 363.0]
    BoundsPointQuery#query.offet: Point2D [x = 0.0, y = 0.0]
```

The scene bounds are different, and I am not sure why. This is mostly just a reference for myself for future work on this (not expecting it to be understood by others).

**Comment by brcolow** [comment]
I am trying to bisect the bad commit, currently I have discovered that it fails as early as v4.0.6-alpha.

**Comment by brcolow** [comment]
Here's the commit that happened to break clicking on a table column header: 369955322314b6f28a3724ca1f4b8b6bcea74ead - https://github.com/TestFX/TestFX/commit/369955322314b6f28a3724ca1f4b8b6bcea74ead - now to figure out why.

**Comment by brcolow** [comment]
I *think* I may know why this is happening. When you set the css `id` of a `TableColumn`, it sets the css id of all the table cells in the first column (including the header, and all of the actual cells in the first column of the table).

8bd3d78b455c56579388b4928329b71efc77ae82 and before has `NodeQueryImpl#tryQuery(..)` that uses Guava:

```java
public <T extends Node> Optional<T> tryQuery() {
    FluentIterable<Node> query = FluentIterable.from(parentNodes);
    return (Optional<T>) query.first();
}
```

This returns `TableColumnHeader[id=tableColumn0, styleClass=column-header table-column]` every time.

Starting with 369955322314b6f28a3724ca1f4b8b6bcea74ead this was changed to:

```java
public <T extends Node> Optional<T> tryQuery() {
    return (Optional<T>) parentNodes.stream().findFirst();
}
```

which happens to return a seemingly random table cell in the first column. This happens to accord with what I was able to reproduce visually.

So part of this has to do with the difference between Guava's `FluentIterable.first()` and Java 8's `Optional.findFirst()`. The other part of this is a broader question about what `NodeQuery.tryQuery` should guarantee as part of its API specification when multiple nodes are matched. Should it just simply rely on the order of nodes as returned by `Node.lookupAll(..)`? If that's the case, then we need to make sure the new `tryQuery` specifically returns the first element of `parentNodes` and not call `findFirst` of `Optional` whose Javadocs say `If the stream has no encounter order, then any element may be returned. `. A more ideal long-term solution is expanding the TestFX API to encompass the hierarchical concept of the scene-graph so that we can say things like "Give me the first matching node that is the highest node in the scene-graph", or "Give me the first matching node that is an ancestor of node A, sibling of node B, etc.`.
